include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -15,30 +17,74 @@ namespace hlsl
 namespace bxdf
 {
 
+// returns unnormalized vector
+// TODO: template these?
+float computeUnnormalizedMicrofacetNormal(bool _refract, float3 V, float3 L, float orientedEta)
+{
+    const float etaFactor = (_refract ? orientedEta : 1.0);
+    const float3 tmpH = V + L * etaFactor;
+    return _refract ? (-tmpH) : tmpH;
+}
+// returns normalized vector, but NaN when 
+float3 computeMicrofacetNormal(bool _refract, float3 V, float3 L, float orientedEta)
+{
+    const float3 H = computeUnnormalizedMicrofacetNormal(_refract,V,L,orientedEta);
+    const float unnormRcpLen = rsqrt(dot(H,H));
+    return H * unnormRcpLen;
+}
+
+// if V and L are on different sides of the surface normal, then their dot product sign bits will differ, hence XOR will yield 1 at last bit
+bool isTransmissionPath(float NdotV, float NdotL)
+{
+    return bool((asuint(NdotV) ^ asuint(NdotL)) & 0x80000000u);
+}
 
 namespace ray_dir_info
 {
 
+#define NBL_CONCEPT_NAME Basic
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)(typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)(U)
+#define NBL_CONCEPT_PARAM_0 (rdirinfo,T)
+#define NBL_CONCEPT_PARAM_1 (N,vector<U, 3>)
+#define NBL_CONCEPT_PARAM_2 (dirDotN,U)
+NBL_CONCEPT_BEGIN(4)
+#define rdirinfo NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define N NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+#define dirDotN NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_2
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.direction), ::nbl::hlsl::is_same_v, vector<U, 3>))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.getDirection()), ::nbl::hlsl::is_same_v, vector<U, 3>))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.transmit()), ::nbl::hlsl::is_same_v, T))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((rdirinfo.reflect(N, dirDotN)), ::nbl::hlsl::is_same_v, T))
+) && nbl::hlsl::is_scalar_v<U>;
+#undef dirDotN
+#undef N
+#undef rdirinfo
+#include <nbl/builtin/hlsl/concepts/__end.hlsl>
+
 // no ray-differentials, nothing
-struct Basic
+template <typename T>
+struct SBasic
 {
-  float3 getDirection() {return direction;}
-
-  Basic transmit()
-  {
-    Basic retval;
-    retval.direction = -direction;
-    return retval;
-  }
-  
-  Basic reflect(const float3 N, const float directionDotN)
-  {
-    Basic retval;
-    retval.direction = nbl::hlsl::reflect(direction,N,directionDotN);
-    return retval;
-  }
-
-  float3 direction;
+    using vector_t = vector<T, 3>;
```

Comment:
I'd call it `vector_type` to keep with stdlib conventions for names of nested typedefs

Reply:
and add a `scalar_type`